sat/cli/bootprep/input/session_template.py

```diff
@@ -95,6 +112,7 @@ def image_is_uuid(self):
         return bool(uuid_regex.fullmatch(self.image))
 
     @cached_property
+    @provides_context('image')
```

Comment:
Still working on understanding this bug and considering your fix and how it interacts with Eli's fix proposed in #37.

As discussed in scrum today, I noticed that your example is creating an image from a base image, but not configuring it, and I noticed that `sat bootprep` does not handle that very well. I filed a [CRAYSAT-1529](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1529) for that separate problem.

Reply:
My original goal with this change was to encourage users to switch to using the `ref_name` to refer to images from the input file in their session templates (as implemented in CRAYSAT-1410 PR #37). If the user is templating their image names in the input file, referring to those images by a non-templated `ref_name` is really the only way to achieve that anyway.

However, you have a valid point here. People may have older input files that use this old way of referencing images from the input file (by their non-templated names) in BOS session templates. If it's easy to continue to support that, we probably should.

I'll just add that this problem appears to be specific to performing a dry-run (or validation) against the input file. If you actually run `sat bootprep` in a non-dry-run mode against such an input file, it will build the IMS image (ignoring CRAYSAT-1529 for now), and then when the session templates are validated, it will find the named image in IMS, and will successfully create the BOS session template. This is not an argument against fixing the dry-run/validation case, but I just wanted to make the distinction clear. I do actually want to expose `validate` as a separate action still, e.g. `sat bootprep validate input.yaml` (see [CRAYSAT-1284](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1284))

With respect to your fix, I am okay with the overall idea. I actually briefly had an implementation similar to this where I would set the `image` variable in the context to a dictionary with just the 'name' key if the image was from the input instance. I think I'm going to do that instead of setting the `image` variable in the context to a `BaseInputImage` object. I think I will also limit the information provided in the `image` context variable to just the 'name' when the IMS record is pulled. We can then consider if more of the info provided by IMS is needed. For now, I think it's better to keep the info provided in that context the same between the two cases.

Such a change will require a little bit of change to Eli's commit added to CRAYSAT-1410, but it should not be too involved. We will basically just need to make sure we handle setting the `image` variable context appropriately in both `InputSessionTemplate` classes' `validate_image_exists` logic.

Reply:
Resolved by e8ee3aad37b9442b4e74e1094a378deeaaeabf99

Reply:
Here's a test showing that this problem is resolved (as long as the image isn't a reference to an unchanged existing IMS image; see CRAYSAT-1529).

First, the input file:

```
ncn-m001:~/haasken/bootprep # cat jstanek-comment.yaml
---
images:
- name: haasken-barebones-image
  base:
    ims:
      name: cray-shasta-csm-sles15sp3-barebones.x86_64-csm-1.2
      type: recipe

session_templates:
- name: compute-{{image.name}}
  image: haasken-barebones-image
  configuration: cos-config-27.06
  bos_parameters:
    boot_sets:
      compute:
        node_roles_groups:
        - "Compute"
```

Then running in dry-run mode:

```
ncn-m001:~/haasken/bootprep # sat bootprep run --dry-run jstanek-comment.yaml
INFO: Validating given input file jstanek-comment.yaml
INFO: Input file successfully validated against schema
INFO: Given input did not define any CFS configurations
WARNING: Found 2 public keys in IMS whose contents match the contents of public key file /root/.ssh/id_rsa.pub. Using the first one, but you may wish to clean up the duplicates.
INFO: Using IMS public key with id 12873d8a-051f-4ef0-9f1e-324a6eb5112c
INFO: Found IMS base for image at index 0: recipe with name=cray-shasta-csm-sles15sp3-barebones.x86_64-csm-1.2
One or more images already exist in IMS with the following names: haasken-barebones-image. Would you like to skip, overwrite, or abort? [skip,overwrite,abort] overwrite
INFO: Images with the following names already exist in IMS and would be overwritten: haasken-barebones-image
INFO: Of the 1 that would be created, 1 have no dependencies and would be created first.
INFO: Would create 1 images.
INFO: Dry run, not creating images.
```

I also ran in non-dry-run mode, and skipped the existing image:

```
ncn-m001:~/haasken/bootprep # sat bootprep run --skip-existing-images jstanek-comment.yaml
INFO: Validating given input file jstanek-comment.yaml
INFO: Input file successfully validated against schema
INFO: Given input did not define any CFS configurations
WARNING: Found 2 public keys in IMS whose contents match the contents of public key file /root/.ssh/id_rsa.pub. Using the first one, but you may wish to clean up the duplicates.
INFO: Using IMS public key with id 12873d8a-051f-4ef0-9f1e-324a6eb5112c
INFO: Found IMS base for image at index 0: recipe with name=cray-shasta-csm-sles15sp3-barebones.x86_64-csm-1.2
INFO: Images with the following names already exist in IMS and will be skipped: haasken-barebones-image
INFO: Any images which depend on these images will also be skipped.
INFO: Of the 0 that will be created, 0 have no dependencies and will be created first.
INFO: Creating 0 images.
INFO: Creating images
INFO: Image creation completed successfully
INFO: Creating 1 BOS session template
ncn-m001:~/haasken/bootprep # cray bos v1 sessiontemplate list | name_contains haasken
[
  {
    "boot_sets": {
      "compute": {
        "etag": "a2c298be8246cdf182a277ea1c61f13c",
        "node_roles_groups": [
          "Compute"
        ],
        "path": "s3://boot-images/be05add1-c3a5-4043-bcbd-08ae3e90b124/manifest.json",
        "rootfs_provider": "cpss3",
        "rootfs_provider_passthrough": "dvs:api-gw-service-nmn.local:300:nmn0",
        "type": "s3"
      }
    },
    "cfs": {
      "configuration": "cos-config-27.06"
    },
    "enable_cfs": true,
    "name": "compute-haasken-barebones-image"
  }
]
```